dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EventDataValuesValidationHook.java

```diff
@@ -0,0 +1,142 @@
+import org.hisp.dhis.tracker.TrackerImportStrategy;
+import org.hisp.dhis.tracker.domain.DataValue;
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.hisp.dhis.tracker.validation.TrackerImportValidationContext;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Enrico Colasante
+ */
+@Component
+public class EventDataValuesValidationHook
+    extends AbstractTrackerDtoValidationHook
+{
+    public EventDataValuesValidationHook( TrackedEntityAttributeService teAttrService )
+    {
+        super( Event.class, TrackerImportStrategy.CREATE_AND_UPDATE, teAttrService );
+    }
+
+    @Override
+    public void validateEvent( ValidationErrorReporter reporter, Event event )
+    {
+        TrackerImportValidationContext context = reporter.getValidationContext();
+
+        for ( DataValue dataValue : event.getDataValues() )
+        {
+            addErrorIf( () -> !isValidDateStringAndNotNull( dataValue.getCreatedAt() ),
+                reporter, TrackerErrorCode.E1300, dataValue.getCreatedAt() );
+            addErrorIf( () -> !isValidDateStringAndNotNull( dataValue.getUpdatedAt() ),
+                reporter, TrackerErrorCode.E1301, dataValue.getUpdatedAt() );
+
+            validateDataElement( reporter, context, dataValue );
+        }
+        validateMandatoryDataValue( reporter, context, event );
+    }
+
+    private void validateDataElement( ValidationErrorReporter reporter, TrackerImportValidationContext ctx,
+        DataValue dataValue )
+    {
+        DataElement dataElement = ctx.getDataElement( dataValue.getDataElement() );
+
+        if ( dataElement == null )
+        {
+            addError( reporter, TrackerErrorCode.E1304, dataValue.getDataElement() );
+        }
+        else
+        {
+            final String status = ValidationUtils.dataValueIsValid( dataValue.getValue(), dataElement );
+
+            if ( status != null )
+            {
+                addError( reporter, TrackerErrorCode.E1302, dataElement.getUid(), status );
+            }
+            else
+            {
+                validateFileNotAlreadyAssigned( reporter, dataValue, dataElement );
+            }
+        }
+    }
+
+    public void validateMandatoryDataValue( ValidationErrorReporter reporter, TrackerImportValidationContext ctx,
+        Event event )
+    {
+        if ( StringUtils.isEmpty( event.getProgramStage() ) )
+            return;
+
+        ProgramStage programStage = ctx.getProgramStage( event.getProgramStage() );
+
+        final Set<ProgramStageDataElement> mandatoryDataElements = programStage.getProgramStageDataElements();
+
+        Set<String> eventDataElements = event.getDataValues().stream()
+            .map( DataValue::getDataElement )
+            .collect( Collectors.toSet() );
+
+        for ( ProgramStageDataElement mandatoryDataElement : mandatoryDataElements )
```

Comment:
This is closely related to Luciano's comment:

Luciano's comment is that we had a previous issue where you can update a data value that is not part of an event/programstage.

This means that if we retrieve the full list of ProgramStageDataElement for a given ProgramStage, all datavalues we valiate needs to have a data element that is part of that list. If it's not part of that list, we are trying to alter data that should not exist.

Since this is only for mandatory data elements, I would say create a new method for validating this, let's say "validateDataValueDataElementIsConnectedToProgramStage" or something :)